metis-media-service/src/main/java/eu/europeana/metis/mediaprocessing/http/AbstractHttpClient.java

```diff
@@ -106,71 +79,139 @@ private void cleanConnections(PoolingHttpClientConnectionManager connectionManag
+    final Optional<String> mimeType = httpResponse.headers().firstValue("Content-Type");
+    final long fileSize = httpResponse.headers().firstValueAsLong("Content-Length").orElse(0);
+
+    // Process the result.
+    final ContentRetriever content = httpResponse.body() == null ?
+        ContentRetriever.forEmptyContent() : httpResponse::body;
+    final R result;
+    try {
+      result = createResult(resourceEntry, actualUri, mimeType.orElse(null),
+              fileSize <= 0 ? null : fileSize, content);
+    } catch (IOException | RuntimeException e) {
+      if (bodyWrapper.isCancelled()) {
+        throw new IOException("The request was aborted: it exceeded the time limit.");
+      }else{
+        throw e;
+      }
+    }
+
+    // If aborted (and createResult did not throw an exception) throw exception anyway.
+    if (bodyWrapper.isCancelled()) {
+      throw new IOException("The request was aborted: it exceeded the time limit.");
+    }
 
-      // Obtain header information.
-      final String mimeType = Optional.ofNullable(responseEntity).map(HttpEntity::getContentType)
-              .orElse(null);
-      final Long fileSize = Optional.ofNullable(responseEntity).map(HttpEntity::getContentLength)
-          .filter(size -> size >= 0).orElse(null);
-      final RedirectLocations redirectUris = context.getRedirectLocations();
-      final URI actualUri = (redirectUris == null || redirectUris.size() == 0) ? httpGet.getUri()
-              : redirectUris.get(redirectUris.size() - 1);
+    // Cancel abort trigger
+    timer.cancel();
+    abortTask.cancel();
 
-      // Process the result.
-      final ContentRetriever content = responseEntity == null ?
-          ContentRetriever.forEmptyContent() : responseEntity::getContent;
-      return createResult(resourceEntry, actualUri, mimeType, fileSize, content);
+    return result;
+  }
 
-    } catch (URISyntaxException e) {
+  private HttpResponse<InputStream> performRedirect(int statusCode, URI location, int redirectsLeft,
+      CancelableBodyWrapper<InputStream> bodyWrapper)
+      throws IOException {
 
-      // Shouldn't really happen.
-      throw new IOException("An unexpected exception occurred.", e);
+    HttpResponse<InputStream> httpResponse = null;
 
-    } catch (IOException e) {
+    while (Status.Family.familyOf(statusCode) == Family.REDIRECTION) {
+      if (redirectsLeft > 0 && location != null) {
+        httpResponse = makeHttpRequest(location, bodyWrapper);
 
-      // If aborted, provide a nicer message. Otherwise, just rethrow.
-      if (httpGet.isAborted()) {
-        throw new IOException("The request was aborted: it exceeded the time limit.", e);
-      }
-      throw e;
+        if(httpResponse != null){
+          statusCode = httpResponse.statusCode();
+          location = httpResponse.headers().firstValue("Location").isEmpty() ?
+              location : location.resolve(httpResponse.headers().firstValue("Location").get());
```

Comment:
Why not better simplify to:
location = httpResponse.headers().firstValue("Location").map(location::resolve).orElse(location);

Reply:
Indeed, it is more simple, will do that